### ISSUE TITLE
Create opensearch.org Maven repository

### ISSUE THREAD

**Issue by dblock** [opensearch-project/opensearch-build#20]

**Comment by mch2** [comment]
Would like to take a stab at this.  We have an issue to do this for lucene as well [here](https://github.com/opensearch-project/OpenSearch/issues/607). As a first step, am thinking of putting together a small cdk project to create the resources.  Will then look at updating infra to publish a nightly.

**Comment by camerski** [comment]
@mch2 has been prototyping some options, and it looks like we have a path forward for automatically publishing everything through Sonatype with no human intervention needed. We probably don't need a self-hosted repository.

**Comment by mch2** [comment]
Updated with a simple configuration to push to our Sonatype repo with ./gradlew publish.  I tested this with a locally running instance of nexus.

I do not think we need [#41] anymore, closing that out.

After this is added to CI, I am thinking we can extend this with a custom script in opensearch-build that builds/publishes each repo we would like to push.

**Comment by mch2** [comment]
We did a successful test push here of only the opensearch repo - https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch.  @camerski is in the process of adding this to the nightly.

**Comment by dblock** [comment]
See https://discuss.opendistrocommunity.dev/t/maven-repository-artifacts-for-plugin-development/6406 for another ask for this.

**Comment by frotsch** [comment]
Pls. also refer to https://github.com/opensearch-project/OpenSearch/issues/939

**Comment by frotsch** [comment]
Why only snapshots and not also the released artifacts?

For plugin developers it's IMHO an absolute must have. Local maven repo does not work well with distributed CI environments etc ...

So I like to ask to publish all the artifacts (releases and snapshots) to maven central like elastic did it/does it.

**Comment by dblock** [comment]
> Why only snapshots and not also the released artifacts?
> 
> For plugin developers it's IMHO an absolute must have. Local maven repo does not work well with distributed CI environments etc ...
> 
> So I like to ask to publish all the artifacts (releases and snapshots) to maven central like elastic did it/does it.

Absolutely. I edited the issue accordingly.

**Comment by frotsch** [comment]
https://discuss.opendistrocommunity.dev/t/how-will-the-opensearch-plugins-be-delivered-in-ga-release/6250

**Comment by dblock** [comment]
**Update**: We still don't have this working end-to-end unfortunately, trying to resolve issues around managing maven credentials across the org. In the meantime the projects continue using [a workaround](https://github.com/opensearch-project/job-scheduler/blob/main/.github/workflows/test-and-build-workflow.yml#L28) of building dependencies into maven local.

**Comment by frotsch** [comment]
Also the clients needs to be published: https://github.com/opensearch-project/OpenSearch/issues/1005

**Comment by rursprung** [comment]
would it be possible to publish this to maven central as well? that'd make live for most of us a lot easier than if we have to use your 3rd party maven repository (companies with their own maven server generally already have maven central set up as a proxy and also for everbody else it's easier if they don't have to manually specify another maven repo in their build scripts).

**Comment by pkalliok-aiven** [comment]
:+1: for having release artifacts in at least sonatype.

I would also like to have the binary distribution packages for plugins, but maybe I'll make another issue out of that.

**Comment by dblock** [comment]
Let's prioritize released 1.0 artifacts if we must make choices.

**Comment by dblock** [comment]
We're trying the following plan:

1. Repository kicks off a GH workflow.
2. Workflow assumes a role to fetch maven credentials from central account via ASM and sets them as an env variable.
3. Workflow fetches pgp subkey to [sign their artifacts](https://docs.gradle.org/current/userguide/signing_plugin.html#using_in_memory_ascii_armored_openpgp_subkeys) from their own account via ASM and sets them as an env variable for sign plugin.
4. Maven artifacts are published to Sonatype Staging repository.
5. Plugin zips are pushed to an S3 bucket living within the repo managed account.
8. After release testing is performed on the bundle, and a build is chosen to be released, a human manually promotes staged artifacts that match that release from Staging to Maven Central.

**Comment by mch2** [comment]
@dblock small update here - we can eliminate step 5 and have each plugin add their zip as a maven publication, we can then pull those from maven when assembling bundles.

**Comment by mch2** [comment]
So these are my ideas on how to fit this all together:

The overall flow:
On-merge to each component repo we rebuild and push a snapshot version to Maven.
Nightly a snapshot bundle is assembled and published to artifacts.opensearch.org.

At release time:
Each component triggers a release build that pushes a staging artifact to Maven.
A release bundle is assembled and published to artifacts.opensearch.org.

Bundle builds will be built from a manifest in the -build repo ([similar to this](https://github.com/opensearch-project/opensearch-build/blob/main/manifests/manifest.linux.x64.example.yml)).
Nightly snapshot bundle builds pick up the latest versions from maven and push to artifacts.opensearch.org.
When we cut a release bundle we will create a new manifest that locks in to specific build versions of each zip artifact.

For component builds/publishing:

Each repository will maintain its own GHA CI scripts within its repository. This is similar to existing ODFE build with some changes. 

1. Common entry-points & templates
We migrate build specific detail from these scripts into common CI agnostic entry-points similar to the integtest.sh proposal (https://github.com/opensearch-project/opensearch-build/issues/58) already in place.  In addition, infra provides a set of common GHA templates for each repository to use in their repositories that invokes these common entry-points.  
Draft:  https://github.com/adnapibar/.github/commit/3c769538536fb5d0c50d6c9a10a17b34b0344746
These additional scripts would be:
* build.sh - wraps all steps required to build the artifact. If the repo is cloned locally this script should execute and produce an artifact without issue.
* publish-snapshot.sh - wraps all steps required to publish a snapshot version to https://aws.oss.sonatype.org/content/repositories/snapshots/
* publish-staging.sh - wraps steps required to publish a staging version to https://aws.oss.sonatype.org/content/repositories/staging.
Triggers would look something like: 
On PR:
The repository runs a precommit workflow, which will build the code and run any relevant tests (via build.sh ). Since this is orchestrated by GitHub Actions, results and logs will be easily visible to the community.
On Merge:
The repository runs a CI workflow that invokes build.sh and then publish-snapshot.sh.
Staging for releases:
A repo admin triggers a release workflow that invokes build.sh, integTest.sh, and publish-staging.sh.

2. Required Reviewers
In order to maintain standards across the project we create a set of required reviewers for .github/workflow files using [CODEOWNERS](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/about-code-owners).

3. Infra Managed maven promotion
There will be a manual promotion by an infra member from staging to maven central after a bundle release.  A lot of heavy integration tests are executed only after a bundle is built, so we cannot promote to maven central until all these tests have been executed. 

4. Plugin zip publishing.
Plugin zips should be added as a maven publication. During bundle creation time we will fetch these artifacts from maven to assemble bundles. We will keep a manifest for snapshot and staging bundle builds to know which version of each plugin zip to include in a bundle.

5. Credential management
In order to publish to maven, each repo needs access to sonatype credentials and signing keys. Publish scripts should assume a role to fetch necessary credentials from Amazon Secrets Manager. This means that each of these repositories should create an account and IAM User that will be granted access to assume this role and fetch these credentials as part of their publish workflow.  An example of this will be provided.  In addition to the Sonatype credentials, a new pgp subkey will be issued and owned by each repo to sign artifacts via ASM.

**Comment by stockholmux** [comment]
@mch2 I assume "GHA" is GitHub Actions?

**Comment by minalsha** [comment]
> @mch2 I assume "GHA" is GitHub Actions?

Yes.

**Comment by frotsch** [comment]
https://github.com/opensearch-project/OpenSearch/issues/1027

**Comment by vrozov** [comment]
I don't see a need to publish *plugin* artifacts to maven on every push to `main`. Will it be sufficient to publish plugin release artifacts? For push, there is an option to publish unreleased artifacts on github if they need to be consumed by other components.

**Comment by vrozov** [comment]
@mch2 How `build.sh`, `publish-snapshot.sh` and `publish-staging.sh` will be used outside of GHA? For example, locally I build differently compared to CI as most of the time I don't need the same full set of verifications while working on my changes before opening PR.

**Comment by mch2** [comment]
> I don't see a need to publish plugin artifacts to maven on every push to main. Will it be sufficient to publish plugin release artifacts? For push, there is an option to publish unreleased artifacts on github if they need to be consumed by other components.

It would make it easier for us when assembling nightly snapshot bundles to use maven to fetch snapshot versions of each plugin zip vs a static url that changes on each build.  We also have cross repo dependencies that are expected to be pulled from a maven repo.  I'd prefer that location be standardized across the project so that any snapshot dependency comes from the same spot.  Given we are going to use Sonatype to promote to central I figured it was a better choice.  I looked at using GH packages for snapshots but there were some limitations - mainly that pulling dependencies locally requires a PAT so they are not truly publicly accessible.

> @mch2 How build.sh, publish-snapshot.sh and publish-staging.sh will be used outside of GHA? For example, locally I build differently compared to CI as most of the time I don't need the same full set of verifications while working on my changes before opening PR.

Have been chewing on this. The intent of the build.sh script is to produce an artifact so that there is no difference between how you'd produce it locally vs from CI.  If those verifications are not required to compile the artifact and/or are too heavy when building locally (like integTest.sh), I'd suggest we pull those steps into a separate validate script. thoughts?  

Also, I don't think we will need the publish scripts in each repo - we will provide a central set of scripts to accomplish those tasks that can be cloned and executed within a workflow and will succeed as long as each repo's IAM user has access to fetch the credentials.  I'd prefer all of that be distributed as a custom action, but I don't think we can share it across the org without publishing to GH marketplace.

**Comment by vrozov** [comment]
>It would make it easier for us when assembling nightly snapshot bundles to use maven to fetch snapshot versions of each plugin zip vs a static url that changes on each build.

While it will make it easier to build, it will make builds unstable and not repeatable as they will be done against moving targets with very limited option to understand what commits are part of the build and which one are not.  This is exactly why maven SNAPSHOT repos are disabled by default.

For assembling snapshot bundles I would consider building every artifact as part of bundle builds itself and recording commit id for each artifact in a bundle manifest file. This will allow to centralize maven publishing and avoid requirements to have access to maven (credentials) for each repo. The bundles builds can be lightweight builds (no tests) as each component of the build will provide guarantee that tip of the `main` is buildable and passes verification/tests.

**Comment by dblock** [comment]
> > It would make it easier for us when assembling nightly snapshot bundles to use maven to fetch snapshot versions of each plugin zip vs a static url that changes on each build.
> 
> While it will make easier to build, it will make builds unstable and not repeatable as they will be done against moving targets with very limited option to understand what commits are part of the build and which one are not. This is exactly why maven SNAPSHOT repos are disabled by default.

Long term we want plugins to have their own destiny and release schedule, though.

> > It would make it easier for us when assembling nightly snapshot bundles to use maven to fetch snapshot versions of each plugin zip vs a static url that changes on each build.
> 
> While it will make easier to build, it will make builds unstable and not repeatable as they will be done against moving targets with very limited option to understand what commits are part of the build and which one are not. This is exactly why maven SNAPSHOT repos are disabled by default.

Long term we want plugins to have their own destiny and release schedule, though.

> For assembling snapshot bundles I would consider building every artifact as part of bundle builds itself and recording commit id for each artifact in a bundle manifest file. This will allow to centralize maven publishing and avoid requirements to have access to maven (credentials) for each repo. The bundles builds can be lightweight builds (no tests) as each component of the build will provide guarantee that tip of the `main` is buildable and passes verification/tests.

This does put the burden of executing artifact build on some centralized process that may/will be different from how each plugin builds its PRs, for example. 

I am not sure whether we can have both worlds, there are pros and cons of either. I think we need CI/CD, first and foremost, and at this point I don't care that much about which way we go.

**Comment by mch2** [comment]
> Long term we want plugins to have their own destiny and release schedule, though.

Right, this was the reason behind the original suggestion having each component put their artifacts somewhere to be picked up in the next bundle build cycle.  However, we currently test bundle releases with integration/bwc/perf tests that are executed after a full bundle is assembled.  So as long as that full bundle exists their releases will be coupled.

> While it will make easier to build, it will make builds unstable and not repeatable as they will be done against moving targets with very limited option to understand what commits are part of the build and which one are not. This is exactly why maven SNAPSHOT repos are disabled by default.

In my mind each zip would include a manifest with the commit from which it was built that would be read and recorded during bundle assembly, but I understand the benefit of avoiding each repo owning signing & publishing its own components. 

I'm not opposed to building on a centralized workflow and I think it gets us up and running faster.  The trade-offs I can think of here:

* We will need to continue to adhere to a standard branching strategy so we know which branch of each plugin repo to check-out, build and publish instead of fetching a version from maven.  We are not only building nightly bundles off of main, we will want to rebuild on changes that need to land in a minor/patch versions of the bundle.

* Publishing of components would happen nightly vs on-merge.  We could technically kick off bundle rebuild on a merge to any repo with a repository_dispatch, but this doesn't seem practical with ~10 repos for core and dashboards.

* As @dblock called out, the build env on which PR's are verified/tested could be different than the bundle build env.  This will need to be standardized with a public container.

**Comment by nknize** [comment]
Some top level questions:

> On-merge to each component repo we rebuild and push a snapshot version to Maven.

Will this scale to `N` repos * `M` merges? Seems like maybe too high fidelity?  Wouldn't nightly snapshots suffice?



> When we cut a release bundle...

Could we move away from "cutting release" bundles?  Could we move toward cutting release candidates who's only difference between "candidate" and "official" is in name only (the three little `rcX` characters)? In this manner, a "release" is not a magical day; it would be the day the last successful `rc` artifact is renamed, "unreleased" flags are flipped on the website and doc pages, and social media announcements are made.

**Comment by vrozov** [comment]
Plugins can still do an independent release/build if necessary, though the exact process of how that can be done depends on where artifacts will be published. Even now plugins can do independent release on github. Problem with the independent release is that it is limited to plugin only and requires uninstall/install of the updated plugin that is quite problematic for existing docker images. We did this couple of times for the security plugin, but afaik those releases were not widely used/downloaded.

Unified branching strategy is nice to have for the centralized build, but not a prerequisite. There should be a way to specify what plugins are part of the build and what branch it should be built from. By default it can be the same/common branch, but there can be exceptions.

Personally I don't see a need to build full bundle on every merge to every repo. Nightly or twice a day builds should be sufficient for SNAPSHOTs. RC and releases should be built on demand.

Github provides necessary means to standardize the build environments. Version of JDK, maven, gradle and node can be agreed upon. I don't see any issue here. Even if plugin uses different version of JDK to build, it must be compiled to a common Java version that core and all plugins agree to support.

**Comment by vrozov** [comment]
> Have been chewing on this. The intent of the build.sh script is to produce an artifact so that there is no difference between how you'd produce it locally vs from CI. The intent of all these scripts is that I could repeat the entire process locally if I chose to. If those verifications are not required to compile the artifact and/or are too heavy when building locally (like integTest.sh), I'd suggest we pull those steps into a separate validate script. thoughts?

Should be a personal preference of developers/maintainers for each repo. I am too used to `mvn ...` to start using `./build.sh` that will be just a shortcut to `mvn` anyway, especially  that I do not want my local build to be the same as CI most of the time I do builds. I prefer PR to do the full validation and limit my local build to compile and affected unit tests only.

**Comment by mch2** [comment]
> Will this scale to N repos * M merges? Seems like maybe too high fidelity? Wouldn't nightly snapshots suffice?

This was referring to snapshot versions of each plugin's maven publications not the bundle builds.  Bundles would be built nightly.

> Could we move away from "cutting release" bundles? Could we move toward cutting release candidates who's only difference between "candidate" and "official" is in name only (the three little rcX characters)? In this manner, a "release" is not a magical day; it would be the day the last successful rc artifact is renamed, "unreleased" flags are flipped on the website and doc pages, and social media announcements are made.

This is ideal.  I would prefer this for bundles and all of the corresponding maven artifacts. Unfortunately for maven artifacts this is more than a rename of a single file bc -rc1 is added to the path and artifact pom files.

> Should be a personal preference of developers/maintainers for each repo. I am too used to mvn ... to start using ./build.sh that will be just a shortcut to mvn anyway, especially that I do not want my local build to be the same as CI most of the time I do builds. I prefer PR to do the full validation and limit my local build to compile and affected unit tests only.

This was so that we could lock to a standard set of GHA templates across the project without needing to know each project's entry-point.  With a centralized build we don't really need this.  Also to reproduce a bundle I simply have to clone a script in -build repo and run it.

We currently have a few limitations we are working around:
1. We do not have a signing mechanism to use with our keys that can be used from GHA. This is a limitation that we are looking to overcome in the near future, but have no answer for today.
2.  The version of Sonatype Nexus we are using does not behave the way we want in that we can't promote from snapshots->staging->central without rebuilds.

Taking this into account here is my recommendation.

1. Nightly - A Central GHA in Opensearch-build repo clones and compiles each artifact.  A bundle manifest is created that includes the commit hash of each zip and a unique build id.  Bundles and maven artifacts are pushed to a public "snapshot" S3 bucket under /<build-id>.  Maven artifacts should not include any qualifier in filename/poms.
2. Test system (currently using an internal Jenkins CI instance) is notified that a new bundle is ready for testing. #111 
3. On completion - Test system triggers a promotion job that fetches all artifacts correlating to the build from "snapshot" S3 buckets, signs and promotes them to "staging" S3 buckets.  This is so anything in /staging is our last known good state.
4. On release day there is no rebuild- we promote the latest "staging" version of bundles to /releases in artifacts.opensearch.org. A separate job fetches all maven artifacts correlating to the bundle version and promotes them to Sonatype as a staging repo, a human logs in and promotes the staging repo to central.

Precommit & on-merge CI will still require some standardization across the project in terms of DCO checks, codecov & other static scans.  It is understood that artifacts produced here will not end up in any shipped bundle.

Longer term - I'd still rather decouple this with a per-repo CI that builds, signs, publishes artifacts and mark which version should be included in a bundle. This depends on a few things:
1.  Each repo needs to be able to run their integ test suite on-merge vs waiting for nightly bundles to be confident in the artifact.
2. A consistent signing mechanism for artifacts usable from GHA.

**Comment by adnapibar** [comment]
> 3. On completion - Test system triggers a promotion job that fetches all artifacts correlating to the build from "snapshot" S3 buckets, signs and promotes them to "staging" S3 buckets.  This is so anything in /staging is our last known good state.

My understanding is snapshots are created from main while releases will be from release branches. So how does this handle that?

**Comment by camerski** [comment]
Snapshots can be created from any branch. In this model, think of a snapshot as just the first step of the build-test-release cycle. Not all snapshots will result in a release, but every release begins with a snapshot.

For example, if we're working on a bugfix that will eventually go into the 1.0.1, we would want daily snapshots generated from the 1.0 branch so that the CI system can run tests every day, and if the tests pass then we have a ready-to-ship 1.0.1 bundle ready for when we decide to release.

**Comment by adnapibar** [comment]
Got it! So is it correct to say that we will build daily bundle snapshots from all these branches e.g main, 1.x, 1.0 ... 2.x, 2.0 ...?

**Comment by camerski** [comment]
I believe that is correct.

**Comment by dblock** [comment]
The additional limitation of the centralized process is that it will force plugins into adopting a consistent branching strategy. Which may be a good thing for now.

**Comment by vrozov** [comment]
IMO, centralized build needs to support building artifacts using git refs (both branches and tags). Each repo should be able to overwrite default branch refs and specify what tag or custom branch should be used for the centralized build.

**Comment by dblock** [comment]
> IMO, centralized build needs to support building artifacts using git refs (both branches and tags). Each repo should be able to overwrite default branch refs and specify what tag or custom branch should be used for the centralized build.

That's fair. As of the time of writing this I should be able to see the complete source code for the bundle build jobs for 1.0, 1.1 and 2.0 and be able to PR into the configuration which ref/branch to use for each component.

**Comment by mch2** [comment]
>> IMO, centralized build needs to support building artifacts using git refs (both branches and tags). Each repo should be able to overwrite default branch refs and specify what tag or custom branch should be used for the centralized build.

> That's fair. As of the time of writing this I should be able to see the complete source code for the bundle build jobs for 1.0, 1.1 and 2.0 and be able to PR into the configuration which ref/branch to use for each component.

We are working on this idea here - #111 to describe a manifest that is used as input to a bundle build and included with each tarball.

**Comment by vrozov** [comment]
#111 talks more about manifest or build metadata output compared to input to the build. For output, IMO, the most convenient format is a single file. For input, I'd suggest to allow each plugin to overwrite default branch/tag used by the build. One possibility is to use build repo branch as the default and have directories named as other components repos with a standard file (for example `.refs`) that allows component to specify git refs to use.

**Comment by mch2** [comment]
1.0 artifacts for core, common-utils, job-scheduler-spi, and notification are now available in [ maven central](https://repo1.maven.org/maven2/org/opensearch/). 

Nightly snapshot builds are pushed [here](https://aws.oss.sonatype.org/content/repositories/snapshots/) by [this](https://github.com/opensearch-project/opensearch-build/blob/main/bundle-workflow/Jenkinsfile) job.

I am in the [process](https://github.com/opensearch-project/opensearch-build/pull/309) of updating the [README](https://github.com/opensearch-project/opensearch-build/blob/main/bundle-workflow/README.md) with information for repo onboarding & details on the bundle build system.

Visibility into the builds can be tracked [here](https://github.com/opensearch-project/opensearch-build/issues/102).

**Comment by newUserForTesting** [comment]
Hello Experts,

Could you please make this also available in maven central: https://github.com/opensearch-project/sql/

Appreciate your help!

**Comment by rursprung** [comment]
from what i can see you only publish the individual JARs & co.
could you please also publish the ZIP files for the actual distribution (incl. the -min distribution!)? that's the one i'd need on maven.

**Comment by dblock** [comment]
Do you need it in a Maven repo, or do you need it "somewhere", and if it's the former, how do you want to use it?

**Comment by rursprung** [comment]
> Do you need it in a Maven repo, or do you need it "somewhere", and if it's the former, how do you want to use it?

in a maven repo. we use gradle to orchestrate our builds, incl. building docker images. so the gradle build pulls in the artifacts (e.g. the ZIP files) from the maven repo and puts it into the build directory, for the Dockerfile it then looks as if the file would be in the same directory (so you can just do `COPY foo.zip /tmp/foo.zip` even though `foo.zip` is not present in your git repo & working tree).

right now i have to manually download the ZIP file and then manually upload it to our internal maven repository, which is very cumbersome since i need to repeat this for every release of every such ZIP file (elasticsearch / opensearch, kibana / opensearch dashboards, plugins, etc.).

**Comment by rursprung** [comment]
is there any indication if this will make it for 1.1 next week?

**Comment by dblock** [comment]
Do you mean zips @rursprung? We haven’t been publishing zips to maven, and I am still not convinced we should, but all the Java artifacts are in AWS OSS Sonatype and will be promoted to maven central. 

 https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#consume-maven-artifacts-in-order

**Comment by rursprung** [comment]
> Do you mean zips @rursprung?

sorry for not being more specific. yes, i meant the distribution ZIPs

> We haven’t been publishing zips to maven, and I am still not convinced we should (maven is really for Java dependencies - who else publishes zips?), but all the Java artifacts are in AWS OSS Sonatype and will be promoted to maven central.
> 
> https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#consume-maven-artifacts-in-order

some examples of others which also publish ZIPs on maven:
* Search Guard: https://docs.search-guard.com/latest/search-guard-versions (all links go to their maven server) and we're actively using that for downloads
* Wildfly: e.g. https://repo1.maven.org/maven2/org/wildfly/wildfly-dist/24.0.1.Final/

maven can be used for things other than java dependencies and is also actively used in such cases.

**Comment by dblock** [comment]
I'd like to know what @nknize @mch2 and @bbarani think?

**Comment by dblock** [comment]
My concerns are:

- The fact that we will no longer have a single authoritative source for artifacts and there will be two ways to get the same thing from two different places.
- Lesser of a concern, the proliferation of snapshot artifacts, times every build.

Note that OpenSearch even inherited [DistributionDownloadPlugin](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/DistributionDownloadPlugin.java) that says that the authors didn't want to put package artifacts into maven. 

Finally, others can also ask for artifacts to be published into <insert my favorite package manager here> and then would we say no?

**Comment by peternied** [comment]
> we use gradle to orchestrate our builds, incl. building docker images.

@rursprung What about a [docker image](https://hub.docker.com/r/opensearchproject/opensearch) instead of a zip file, if there was a `-min` version of OpenSearch would that be more in alignment with your process?
